parachain/src/wasm_executor/validation_host.rs

```diff
@@ -293,23 +326,23 @@ impl ValidationHost {
 			let mut message_data: &[u8] = message_data;
 			let header = ValidationResultHeader::decode(&mut header_buf).unwrap();
 			match header {
-				ValidationResultHeader::Ok { result, egress_message_count, up_message_count } => {
-					for _ in 0 .. egress_message_count {
-						let message = IncomingMessage::decode(&mut message_data).unwrap();
-						let message_ref = MessageRef {
-							target: message.source,
-							data: &message.data,
-						};
-						externalities.post_message(message_ref)?;
-					}
-					for _ in 0 .. up_message_count {
-						let message = UpwardMessage::decode(&mut message_data).unwrap();
-						let message_ref = UpwardMessageRef {
-							origin: message.origin,
-							data: &message.data,
-						};
-						externalities.post_upward_message(message_ref)?;
-					}
+				ValidationResultHeader::Ok(result) => {
+					let egress = Vec::<TargetedMessage>::decode(&mut message_data)
+						.map_err(|e|
+							Error::External(
+								format!("Could not decode egress messages: {}", e.what())
+							)
+						)?;
```

Comment:
I struggle to understand how this egress gets written, is it validation_code responsibility to return those two vectors ? or is it written by the worker somehow at the line 158 ?
```rust
message_data[0..e_len].copy_from_slice(egress_data);
```

 I see WorkerExternalities store them in Vec now

Reply:
First it gets written here to a `Vec<u8>` in the validation host: https://github.com/paritytech/polkadot/pull/665/files#diff-6f0269e206acb5ff83e3c41e68c2d451R55

After the validation is finished, we write the data here into the shared memory: https://github.com/paritytech/polkadot/pull/665/files#diff-6f0269e206acb5ff83e3c41e68c2d451R158

And this code here, is the "other" side of the shared memory in the polkadot node that reads the result.

Reply:
ok got it.

but in the shared memory egress are written as an array not a vec: 
https://github.com/paritytech/polkadot/pull/665/files#diff-6f0269e206acb5ff83e3c41e68c2d451R158
```rust
message_data[0..e_len].copy_from_slice(egress_data);
```
And here it is decoded as a vec. it seems to miss the size of the vec for decoding.

Reply:
We copy the whole encoded buffer into the shared memory (The encoded buffer is the correct SCALE encoded vec of egress messages). So, we don't miss to write the length, it's already in the data.

Reply:
oh yes indeed